### PR TITLE
HIP_ACO successor list refactor

### PR DESCRIPTION
**Pull request by balinck**
Summary of changes:
  * Information about successors is consolidated into scsrs_, latencies_, predOrder_ for instructions being sent to device.
  * Removed __device__ flag from many functions relying on the use of GraphEdge objects.
  * Added some GraphNode functionality to SchedInstruction for use on the device:
    * SchedInstruction::GetScsr() used in place of GraphNode::GetFrstScsr()/GetNxtScsr()
    * SchedInstruction::GetScsrCnt_() used in place of GraphNode::GetScsrCnt()
  * GraphNode::IsLeaf() returns the value of dev_IsLeaf when called on the device.
  * ACO scheduler and the list scheduler use the new SchedInstruction methods to traverse successors when on the device.

### CODE REVIEW

File: lib/Scheduler/sched_basic_data.hip.cpp [hunk 1]
```diff
@@ -428,11 +431,33 @@ int SchedInstruction::GetNodeID() const { return nodeID_; }
...
+// TODO(bruce) consider storing this value in SetupForDevice instead of recomputing
 __host__ __device__
-int SchedInstruction::GetLtncySum() const { return GetScsrLblSum(); }
+int SchedInstruction::GetLtncySum() const {
+  #ifdef __HIP_DEVICE_COMPILE__
+  int sum = 0;
+  for (int i = 0; i < scsrCnt_; i++) {
+    sum += latencies_[i];
+  }
+  #else
```
Comment (JoshHuttonCode) [inline lib/Scheduler/sched_basic_data.hip.cpp]: Unless I misunderstand, I think this should be returning sum.

File: lib/Scheduler/sched_basic_data.hip.cpp [hunk 2]
```diff
@@ -428,11 +431,33 @@ int SchedInstruction::GetNodeID() const { return nodeID_; }
...
 __host__ __device__
-int SchedInstruction::GetMaxLtncy() const { return GetMaxEdgeLabel(); }
+int SchedInstruction::GetMaxLtncy() const { 
+  #ifdef __HIP_DEVICE_COMPILE__
+  int max = 0;
+  for (int i = 0; i < scsrCnt_; i++) {
+    if (latencies_[i] > max) {
+      max = latencies_[i];
+    }
+  }
```
Comment (JoshHuttonCode) [inline lib/Scheduler/sched_basic_data.hip.cpp:457]: Similarly, I think this should return max.

File: lib/Scheduler/sched_basic_data.hip.cpp [hunk 3]
```diff
@@ -443,11 +468,11 @@ int16_t SchedInstruction::GetLastUseCnt() {
 #endif
 }
 
-__host__ __device__
+__host__
 SchedInstruction *SchedInstruction::GetFrstPrdcsr(InstCount *scsrNum,
                                                   UDT_GLABEL *ltncy,
                                                   DependenceType *depType,
-						  InstCount *toNodeNum) {
+						                                      InstCount *toNodeNum) {
```
Comment (JoshHuttonCode) [inline lib/Scheduler/sched_basic_data.hip.cpp]: It looks like there's a huge space here.

File: lib/Scheduler/sched_basic_data.hip.cpp [hunk 4]
```diff
@@ -462,11 +487,11 @@ SchedInstruction *SchedInstruction::GetFrstPrdcsr(InstCount *scsrNum,
   return (SchedInstruction *)(nodes_[edge->from]);
 }
 
-__host__ __device__
+__host__
 SchedInstruction *SchedInstruction::GetNxtPrdcsr(InstCount *scsrNum,
                                                  UDT_GLABEL *ltncy,
                                                  DependenceType *depType,
-						 InstCount *toNodeNum) {
+						                                     InstCount *toNodeNum) {
```
Comment (JoshHuttonCode) [inline lib/Scheduler/sched_basic_data.hip.cpp]: Space.

File: lib/Scheduler/sched_basic_data.hip.cpp [hunk 5]
```diff
@@ -481,11 +506,38 @@ SchedInstruction *SchedInstruction::GetNxtPrdcsr(InstCount *scsrNum,
...
+  return (SchedInstruction *) nodes_[scsrs_[scsrNum]];
+
+}
+
+__host__
 SchedInstruction *SchedInstruction::GetFrstScsr(InstCount *prdcsrNum,
                                                 UDT_GLABEL *ltncy,
                                                 DependenceType *depType,
-						InstCount *toNodeNum,
+						                                    InstCount *toNodeNum,
```
Comment (JoshHuttonCode) [inline lib/Scheduler/sched_basic_data.hip.cpp]: space

File: lib/Scheduler/sched_basic_data.hip.cpp [hunk 6]
```diff
@@ -503,11 +555,11 @@ SchedInstruction *SchedInstruction::GetFrstScsr(InstCount *prdcsrNum,
   return (SchedInstruction *)(nodes_[edge->to]);
 }
 
-__host__ __device__
+__host__
 SchedInstruction *SchedInstruction::GetNxtScsr(InstCount *prdcsrNum,
                                                UDT_GLABEL *ltncy,
                                                DependenceType *depType,
-					       InstCount *toNodeNum,
+					                                     InstCount *toNodeNum,
```
Comment (JoshHuttonCode) [inline lib/Scheduler/sched_basic_data.hip.cpp]: spaaaaaaaaaaaace

File: lib/Scheduler/aco.hip.cpp [hunk 7]
```diff
@@ -587,7 +586,7 @@ InstSchedule *ACOScheduler::FindOneSchedule(InstCount RPTarget,
         printf("cyclenum: %d Close to RP Target: %s\n", dev_crntCycleNum_[GLOBALTID], closeToRPTarget ? "true" : "false");
       }
       #endif
-
+      int totalstalls__ = schedule->getTotalStalls();
```
Comment (JoshHuttonCode) [inline lib/Scheduler/aco.hip.cpp]: Why was this added?

**Review by JoshHuttonCode — CHANGES_REQUESTED** [review]
A few legitimate questions and concerns, as well as some random formatting things.

**Comment by balinck** [comment]
The whitespace problems should now be fixed. GetLtncySum() and GetLtncyMax() properly return their results now. I'm not sure why the line in aco.hip.cpp was added, but it doesn't seem to do anything, so I've removed it.

**Review by JoshHuttonCode — APPROVED** [review]
Looks good to me.
